### ISSUE TITLE
First save of a file takes too long

### ISSUE THREAD

**Issue by amol9** [Microsoft/vscode#21546]
- VSCode Version: 1.9.1
- OS Version: Windows 10 (Version 1511, Build 10586.545)

Steps to Reproduce:

1. Create a new file.
2. Type some code in it.
3. Save the file.

It takes 5 seconds on an average for the file name to appear on the tab and syntax highlighting to show if the language was not selected already. I have observed this behavior for about a month now. It is really annoying.

I tried the same steps with Notepad++, it saves and highlights in an instant. So, it's not a performance issue of my machine / OS.

Thanks.

**Comment by bpasero** [comment]
I wonder if this issue is being caused by an installed extension. Can you try to run VS Code without extensions? From the command line (NOT the integrated terminal in Code), execute: `code --disable-extensions` and try your steps again to see if it reproduces. If you see it is an issue with the extension, please file it against the extension repository itself.

**Comment by amol9** [comment]
I tried to reproduce the issue by running VS Code without extensions using the --disable-extensions command line switch.

Unfortunately, the issue is still reproducible. So, it does not seem to be an issue with one of the extensions.

And, things got a little interesting. I noticed that whenever I just typed the word "test" and then a blank new line after it, the issue disappeared, i.e., the file was saved in an instant.

I tried different combinations. (>>> and <<< mark the beginning and end of file, not part of the typed text in a file).

>>>
import os
test

<<<
result: no delay

>>>
import os
<<<
result: delay

>>>
import os
test
import sys

<<<
result: delay

>>>
import os

class Test:
    pass

<<<
result: delay

>>>
import os

class Test:
    pass

test

<<<
result: no delay

>>>
test
<<<
result: delay

>>>
test

<<<
result: delay

I think it has something to do with parsing that is done at the time of saving. And syntax errors seem to affect it.

**Comment by amol9** [comment]
Adding some more info.

I tried saving a plain text file with extension, .txt. Still, there is similar delay of about 5 seconds.

**Comment by bpasero** [comment]
Please see if you can find any logged output that could help us find the source of the issue:
* run `code --verbose` from the command line and attach the output
* in Code select **Help | Toggle Developer Tools** and select the **Console**
* attach any output you see in the **Console** you can find

Thanks!

**Comment by amol9** [comment]
Here's the output from the command line after running with --verbose switch:
<pre>

[main 5:05:28 PM] Starting VS Code in verbose mode
[main 5:05:29 PM] from: c:\Program Files (x86)\Microsoft VS Code\resources\app
[main 5:05:29 PM] args: { _: [],
  help: false,
  h: false,
  version: false,
  v: false,
  wait: false,
  w: false,
  diff: false,
  d: false,
  goto: false,
  g: false,
  'new-window': false,
  n: false,
  'new-window-if-not-first': false,
  'reuse-window': false,
  r: false,
  performance: false,
  p: false,
  verbose: true,
  logExtensionHostCommunication: false,
  'disable-extensions': false,
  disableExtensions: false,
  'list-extensions': false,
  'show-versions': false,
  nolazy: false }
[5016:0302/170553:INFO:CONSOLE(70)] "Ctrl+Alt+ keybindings should not be used by default under Windows. Offender: ", source: file:///C:/Progr
am Files (x86)/Microsoft VS Code/resources/app/out/vs/workbench/electron-browser/workbench.main.js (70)
[main 5:05:54 PM] IPC#vscode-workbenchLoaded
</pre>

In Developer Tools > Console, only this warning showed (before i created a new file and saved it):
<pre>
Ctrl+Alt+ keybindings should not be used by default under Windows. Offender:  2595  for  PowerShell.ExpandAlias
</pre>

After creating a new file and saving it, there was no new information in dev tools console.

**Comment by bpasero** [comment]
Would be good to find out where time is spend via a CPU profile. 

To create a CPU profile using developer tools in VS Code:
* select Help > Toggle Developer Tools
* click on Profiles
* click the start button
* do whatever you need to do to get the CPU spin high
* wait for some time (10-20 seconds should be enough)
* stop the profiling 
* click the save button and attach the output file to this issue

**Comment by amol9** [comment]
I am attaching the output.

Steps:
1. start profile
2. create a new file
3. save it
4. wait for it to save (for the filename to show on the tab)
5. stop profile
6. save profile
[CPU-20170302T172647.cpuprofile.txt](https://github.com/Microsoft/vscode/files/813699/CPU-20170302T172647.cpuprofile.txt)

**Comment by bpasero** [comment]
@amol9 most of the time is spend in the dialog to pick the filename? This is not about saving a file that exists but a new file? Could it be your OS takes long to open that native file dialog?

**Comment by amol9** [comment]
The file dialog shows as soon as I hit Ctrl+S. Then, I type the name of the file and hit enter. The delay comes after the file save dialog is dismissed.

**Comment by bpasero** [comment]
@amol9 from the profile it looks like under 1s is spend in opening that file. What did you experience how long does it take for you?

**Comment by bpasero** [comment]
Actually my question is not a good one, you already mentioned that it depends on the contents of the file, right?

I have another question, does it reproduce from our nightly insiders build?  You can give our preview releases a try from: http://code.visualstudio.com/Download#insiders

**Comment by amol9** [comment]
I am updating the issue after a long time. I have some more info.

I noticed that there was a similar delay of about 5s while saving documents in notepad as well. So, I thought, may be it's an issue with some windows component which is used by both notepad and vs code. I had not updated my system for a few months. So, I tried to update windows.

All pending updates, except, the anniversary update of windows 10 (build 1607) installed successfully. The b1607 update is huge, it took a long time to download and failed to install after downloading.

After the update, I tested for this issue again. Also, I tested with various other microsoft products to see if they have a similar issue. Also, vs code automatically for updated to version 1.10.1 just before I updated windows.

Here are my findings:
vs code : no delay in saving
notepad : delay
word : no delay
wordpad : no delay

So, the issue seems to have disappeared for vs code. Although, I have only tested it briefly with 4-5 files.
